{
    "version": "https://jsonfeed.org/version/1",
    "title": "Hexo • All posts by \"java微服务\" category",
    "description": "",
    "home_page_url": "https://sun-spots.github.io",
    "items": [
        {
            "id": "https://sun-spots.github.io/2022/04/10/java%E5%BE%AE%E6%9C%8D%E5%8A%A1/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/",
            "url": "https://sun-spots.github.io/2022/04/10/java%E5%BE%AE%E6%9C%8D%E5%8A%A1/%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86/",
            "title": "java异常基础知识",
            "date_published": "2022-04-10T12:05:00.000Z",
            "content_html": "<h2 id=\"异常\"><a class=\"markdownIt-Anchor\" href=\"#异常\"></a> 异常</h2>\n<p>异常的概念：异常是造成程序中断的一种指令流，如果不对异常进行处理则可以导致程序的中断运行，造成不必要的损失，所以在程序设计中必须考虑各种异常的发生，并正确的做好相应的处理，这样保证程序的正常运行。</p>\n<p><a href=\"https://imgtu.com/i/LeR760\"><img data-src=\"https://s1.ax1x.com/2022/04/12/LeR760.png\" alt=\"LeR760.png\" /></a></p>\n<figure class=\"highlight plaintext\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br><span class=\"line\">7</span><br><span class=\"line\">8</span><br><span class=\"line\">9</span><br><span class=\"line\">10</span><br><span class=\"line\">11</span><br><span class=\"line\">12</span><br><span class=\"line\">13</span><br><span class=\"line\">14</span><br><span class=\"line\">15</span><br><span class=\"line\">16</span><br><span class=\"line\">17</span><br><span class=\"line\">18</span><br><span class=\"line\">19</span><br><span class=\"line\">20</span><br><span class=\"line\">21</span><br><span class=\"line\">22</span><br></pre></td><td class=\"code\"><pre><span class=\"line\">**java的异常处理机制**：Object是java中任何类的超类。Object下有Throwable（可抛出的），异常对都是派生于Throwable类的一个实例。当java中的内置异常类无法满足用户需要时，可以继承异常类进行编写自己的异常类库。Throwable下有两个分支：Error（不可处理，直接退出JVM）和Exception（可处理的）</span><br><span class=\"line\">Exception下有两个分支：</span><br><span class=\"line\">**编译时异常**（受检异常：CheckedException、受控异常）：</span><br><span class=\"line\">要求程序员在编写程序阶段必须预先对这些异常进行处理，如果不处理编译器则会报错。包括Exception，但不包括其子类RuntimeException</span><br><span class=\"line\">**运行时异常**（未受检异常：UnCheckedException、非受控异常）：在编写程序阶段程序员可以预先处理，也可以不管，当程序出现错误时，可以对出现的异常进行检测或者抛出。</span><br><span class=\"line\">包括：RuntimeException及其子类</span><br><span class=\"line\"></span><br><span class=\"line\">编译时异常和运行时异常，都是发生在运行阶段。编译阶段异常是不会发生的。</span><br><span class=\"line\">所有异常都是在运行阶段发生的。因为只有程序运行阶段才</span><br><span class=\"line\">可以new对象。</span><br><span class=\"line\">因为异常的发生就是new异常对象。</span><br><span class=\"line\"></span><br><span class=\"line\">编译时异常和运行时异常的区别？</span><br><span class=\"line\">编译时异常一般发生的概率比较高。</span><br><span class=\"line\">运行时异常一般发生的概率比较低。</span><br><span class=\"line\"></span><br><span class=\"line\">Java语言中对异常的处理包括两种方式：</span><br><span class=\"line\">第一种方式：在方法声明的位置上，使用throws关键字，抛给上一级。谁调用我，我就抛给谁。</span><br><span class=\"line\"></span><br><span class=\"line\">第二种方式：使用try..catch语句进行异常的捕捉。这件事发生了，谁也不知道，因为我给抓住了。</span><br><span class=\"line\"></span><br><span class=\"line\">注意：Java中异常发生之后如果一直上抛，最终抛给了main方法，main方法继续向上抛，抛给了调用者JVM，JVM知道这个异常发生，只有一个结果。终止java程序的执行。</span><br></pre></td></tr></table></figure>\n<p><strong>异常的捕获机制：</strong></p>\n<figure class=\"highlight java\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br><span class=\"line\">7</span><br><span class=\"line\">8</span><br><span class=\"line\">9</span><br><span class=\"line\">10</span><br><span class=\"line\">11</span><br><span class=\"line\">12</span><br><span class=\"line\">13</span><br></pre></td><td class=\"code\"><pre><span class=\"line\"><span class=\"keyword\">try</span>&#123;</span><br><span class=\"line\">\t<span class=\"comment\">//有可能产生异常的代码</span></span><br><span class=\"line\">&#125;</span><br><span class=\"line\"><span class=\"keyword\">catch</span>( 异常类  异常对象)&#123;</span><br><span class=\"line\">\t<span class=\"comment\">//捕捉异常信息</span></span><br><span class=\"line\">&#125;</span><br><span class=\"line\"><span class=\"keyword\">catch</span>(异常类  异常对象)&#123;</span><br><span class=\"line\">\t<span class=\"comment\">//</span></span><br><span class=\"line\">&#125;</span><br><span class=\"line\">..... <span class=\"comment\">//至少匹配一个catch</span></span><br><span class=\"line\"><span class=\"keyword\">finally</span>&#123;</span><br><span class=\"line\">\t<span class=\"comment\">//不管有没有异常都会被执行到的程序</span></span><br><span class=\"line\">&#125;</span><br></pre></td></tr></table></figure>\n<p><strong>异常的继承关系分析</strong></p>\n<figure class=\"highlight java\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br></pre></td><td class=\"code\"><pre><span class=\"line\">Throwable:</span><br><span class=\"line\">子类：Exception/Error</span><br><span class=\"line\">​\tException的分类： RuntimeException/Exception、</span><br><span class=\"line\">​\tRuntimeException:运行时异常，如果以后碰到RuntimeException，可以不用显式去处理.交给jvm去处理.</span><br><span class=\"line\">​\tException:必须进行显式的处理</span><br><span class=\"line\">既然Throwable是Exception的父类，所有异常是不是可以用Throwable进行接收<span class=\"comment\">//不可以的,因为Throwable也是Error的父类,异常的父类都是Exception</span></span><br></pre></td></tr></table></figure>\n<figure class=\"highlight java\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br><span class=\"line\">7</span><br><span class=\"line\">8</span><br><span class=\"line\">9</span><br><span class=\"line\">10</span><br><span class=\"line\">11</span><br><span class=\"line\">12</span><br><span class=\"line\">13</span><br><span class=\"line\">14</span><br><span class=\"line\">15</span><br><span class=\"line\">16</span><br><span class=\"line\">17</span><br><span class=\"line\">18</span><br><span class=\"line\">19</span><br></pre></td><td class=\"code\"><pre><span class=\"line\"><span class=\"keyword\">import</span> java.io.IOException;</span><br><span class=\"line\"></span><br><span class=\"line\">    <span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">Test</span> </span>&#123;</span><br><span class=\"line\">       <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">        IOException ioException = <span class=\"keyword\">new</span> IOException(<span class=\"string\">&quot;手动new出异常&quot;</span>);</span><br><span class=\"line\">        System.out.println(<span class=\"string\">&quot;Hello World&quot;</span>);</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">运行结果：Hello World</span><br><span class=\"line\">注意：自己手动用<span class=\"keyword\">new</span>动态申请出异常跟正常代码一样,不会影响下面代码的执行。非手动<span class=\"keyword\">new</span>出异常则底层会帮我们产生相应的异常对象,然后抛给调用者,如果调用者不处理,则运行会报错将会终止运行,之后的代码就不会执行。</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">Test</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">   \t<span class=\"comment\">//这里底层会产生java.lang.ArithmeticException,然后抛给main方法这个调用者</span></span><br><span class=\"line\">   \t<span class=\"keyword\">int</span> i = <span class=\"number\">1</span>/<span class=\"number\">0</span>; </span><br><span class=\"line\">   \t<span class=\"comment\">//因为这里没有处理java.lang.ArithmeticException,所以抛给JVM,结果就是直接报错,下面的代码不会运行</span></span><br><span class=\"line\">   \tSystem.out.println(<span class=\"string\">&quot;Hello World&quot;</span>);</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">运行结果：运行结果:Exception in thread <span class=\"string\">&quot;main&quot;</span> java.lang.ArithmeticException: / by zero</span><br></pre></td></tr></table></figure>\n<p><strong>获取异常的两个非常重要的方法</strong></p>\n<figure class=\"highlight java\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br><span class=\"line\">7</span><br><span class=\"line\">8</span><br><span class=\"line\">9</span><br><span class=\"line\">10</span><br><span class=\"line\">11</span><br><span class=\"line\">12</span><br><span class=\"line\">13</span><br><span class=\"line\">14</span><br><span class=\"line\">15</span><br><span class=\"line\">16</span><br><span class=\"line\">17</span><br></pre></td><td class=\"code\"><pre><span class=\"line\">异常对象有两个非常重要的方法：</span><br><span class=\"line\">-    获取异常简单的描述信息：String msg = exception.getMessage();</span><br><span class=\"line\">-    打印异常追踪的堆栈信息：exception.printStackTrace();</span><br><span class=\"line\">-    获取造成异常描述信息：exception.getcause();</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest08</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">       <span class=\"comment\">// 这里只是为了测试getMessage()方法和printStackTrace()方法。</span></span><br><span class=\"line\">       <span class=\"comment\">// 这里只是new了异常对象，但是没有将异常对象抛出。JVM会认为这是一个普通的java对象。</span></span><br><span class=\"line\">       NullPointerException e = <span class=\"keyword\">new</span> NullPointerException(<span class=\"string\">&quot;空指针异常fdsafdsafdsafds&quot;</span>);</span><br><span class=\"line\">       <span class=\"comment\">// 获取异常简单描述信息：这个信息实际上就是构造方法上面String参数。</span></span><br><span class=\"line\">       String msg = e.getMessage(); <span class=\"comment\">//空指针异常fdsafdsafdsafds</span></span><br><span class=\"line\">       System.out.println(msg);</span><br><span class=\"line\">       <span class=\"comment\">// 打印异常堆栈信息</span></span><br><span class=\"line\">       e.printStackTrace();</span><br><span class=\"line\">       System.out.println(<span class=\"string\">&quot;Hello World!&quot;</span>);</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br></pre></td></tr></table></figure>\n<p><strong>关于try…catch中的finally子句</strong><br />\n在finally子句中的代码是最后执行的，并且是一定会执行的，即使try语句块中的代码出现了异常。<br />\nfinally子句必须和try一起出现，不能单独编写。<br />\nfinally语句通常使用在哪些情况下呢？<br />\n通常在finally语句块中完成资源的释放/关闭。<br />\n因为finally中的代码比较有保障。<br />\n即使try语句块中的代码出现异常，finally中代码也会正常执行。</p>\n<figure class=\"highlight java\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br><span class=\"line\">7</span><br><span class=\"line\">8</span><br><span class=\"line\">9</span><br><span class=\"line\">10</span><br><span class=\"line\">11</span><br><span class=\"line\">12</span><br><span class=\"line\">13</span><br><span class=\"line\">14</span><br><span class=\"line\">15</span><br><span class=\"line\">16</span><br><span class=\"line\">17</span><br><span class=\"line\">18</span><br><span class=\"line\">19</span><br><span class=\"line\">20</span><br><span class=\"line\">21</span><br><span class=\"line\">22</span><br><span class=\"line\">23</span><br><span class=\"line\">24</span><br><span class=\"line\">25</span><br><span class=\"line\">26</span><br><span class=\"line\">27</span><br><span class=\"line\">28</span><br><span class=\"line\">29</span><br><span class=\"line\">30</span><br><span class=\"line\">31</span><br><span class=\"line\">32</span><br><span class=\"line\">33</span><br><span class=\"line\">34</span><br><span class=\"line\">35</span><br><span class=\"line\">36</span><br><span class=\"line\">37</span><br><span class=\"line\">38</span><br><span class=\"line\">39</span><br></pre></td><td class=\"code\"><pre><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest10</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">       FileInputStream fis = <span class=\"keyword\">null</span>; <span class=\"comment\">// 声明位置放到try外面。这样在finally中才能用。</span></span><br><span class=\"line\">       <span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">           <span class=\"comment\">// 创建输入流对象</span></span><br><span class=\"line\">           fis = <span class=\"keyword\">new</span> FileInputStream(<span class=\"string\">&quot;D:\\\\course\\\\02-JavaSE\\\\document\\\\JavaSE进阶讲义\\\\JavaSE进阶-01-面向对象.pdf&quot;</span>);</span><br><span class=\"line\">           <span class=\"comment\">// 开始读文件....</span></span><br><span class=\"line\"></span><br><span class=\"line\">           String s = <span class=\"keyword\">null</span>;</span><br><span class=\"line\">           <span class=\"comment\">// 这里一定会出现空指针异常！</span></span><br><span class=\"line\">           s.toString();</span><br><span class=\"line\">           System.out.println(<span class=\"string\">&quot;hello world!&quot;</span>);</span><br><span class=\"line\">           <span class=\"comment\">// 流使用完需要关闭，因为流是占用资源的。</span></span><br><span class=\"line\">           <span class=\"comment\">// 即使以上程序出现异常，流也必须要关闭！</span></span><br><span class=\"line\">           <span class=\"comment\">// 放在这里有可能流关不了。</span></span><br><span class=\"line\">           <span class=\"comment\">//fis.close();</span></span><br><span class=\"line\">       &#125; <span class=\"keyword\">catch</span> (FileNotFoundException e) &#123;</span><br><span class=\"line\">           e.printStackTrace();</span><br><span class=\"line\">       &#125; <span class=\"keyword\">catch</span>(IOException e)&#123;</span><br><span class=\"line\">           e.printStackTrace();</span><br><span class=\"line\">       &#125; <span class=\"keyword\">catch</span>(NullPointerException e) &#123;</span><br><span class=\"line\">           e.printStackTrace();</span><br><span class=\"line\">       &#125; <span class=\"keyword\">finally</span> &#123;</span><br><span class=\"line\">           System.out.println(<span class=\"string\">&quot;hello 浩克！&quot;</span>);</span><br><span class=\"line\">           <span class=\"comment\">// 流的关闭放在这里比较保险。</span></span><br><span class=\"line\">           <span class=\"comment\">// finally中的代码是一定会执行的。</span></span><br><span class=\"line\">           <span class=\"comment\">// 即使try中出现了异常！</span></span><br><span class=\"line\">           <span class=\"keyword\">if</span> (fis != <span class=\"keyword\">null</span>) &#123; <span class=\"comment\">// 避免空指针异常！</span></span><br><span class=\"line\">               <span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">                   <span class=\"comment\">// close()方法有异常，采用捕捉的方式。</span></span><br><span class=\"line\">                   fis.close();</span><br><span class=\"line\">               &#125; <span class=\"keyword\">catch</span> (IOException e) &#123;</span><br><span class=\"line\">                   e.printStackTrace();</span><br><span class=\"line\">               &#125;</span><br><span class=\"line\">           &#125;</span><br><span class=\"line\">       &#125;</span><br><span class=\"line\">       System.out.println(<span class=\"string\">&quot;hello kitty!&quot;</span>);</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br></pre></td></tr></table></figure>\n<figure class=\"highlight java\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br><span class=\"line\">7</span><br><span class=\"line\">8</span><br><span class=\"line\">9</span><br><span class=\"line\">10</span><br><span class=\"line\">11</span><br><span class=\"line\">12</span><br><span class=\"line\">13</span><br><span class=\"line\">14</span><br><span class=\"line\">15</span><br><span class=\"line\">16</span><br><span class=\"line\">17</span><br><span class=\"line\">18</span><br><span class=\"line\">19</span><br><span class=\"line\">20</span><br><span class=\"line\">21</span><br><span class=\"line\">22</span><br><span class=\"line\">23</span><br><span class=\"line\">24</span><br><span class=\"line\">25</span><br><span class=\"line\">26</span><br><span class=\"line\">27</span><br><span class=\"line\">28</span><br><span class=\"line\">29</span><br><span class=\"line\">30</span><br><span class=\"line\">31</span><br><span class=\"line\">32</span><br><span class=\"line\">33</span><br><span class=\"line\">34</span><br></pre></td><td class=\"code\"><pre><span class=\"line\"><span class=\"keyword\">finally</span>面试题</span><br><span class=\"line\"><span class=\"keyword\">package</span> com.yxj.javase.exception;</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest13</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">       <span class=\"keyword\">int</span> result = m();</span><br><span class=\"line\">       System.out.println(result); <span class=\"comment\">//100</span></span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">   <span class=\"comment\">/*</span></span><br><span class=\"line\"><span class=\"comment\">   java语法规则（有一些规则是不能破坏的，一旦这么说了，就必须这么做！）：</span></span><br><span class=\"line\"><span class=\"comment\">       java中有一条这样的规则：</span></span><br><span class=\"line\"><span class=\"comment\">           方法体中的代码必须遵循自上而下顺序依次逐行执行（亘古不变的语法！）</span></span><br><span class=\"line\"><span class=\"comment\">       java中还有一条语法规则：</span></span><br><span class=\"line\"><span class=\"comment\">           return语句一旦执行，整个方法必须结束（亘古不变的语法！）</span></span><br><span class=\"line\"><span class=\"comment\">    */</span></span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">int</span> <span class=\"title\">m</span><span class=\"params\">()</span></span>&#123;</span><br><span class=\"line\">       <span class=\"keyword\">int</span> i = <span class=\"number\">100</span>;</span><br><span class=\"line\">       <span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">           <span class=\"comment\">// 这行代码出现在int i = 100;的下面，所以最终结果必须是返回100</span></span><br><span class=\"line\">           <span class=\"comment\">// return语句还必须保证是最后执行的。一旦执行，整个方法结束。</span></span><br><span class=\"line\">           <span class=\"keyword\">return</span> i;</span><br><span class=\"line\">       &#125; <span class=\"keyword\">finally</span> &#123;</span><br><span class=\"line\">           i++;</span><br><span class=\"line\">       &#125;</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\"><span class=\"comment\">/*</span></span><br><span class=\"line\"><span class=\"comment\">反编译之后的效果</span></span><br><span class=\"line\"><span class=\"comment\">public static int m()&#123;</span></span><br><span class=\"line\"><span class=\"comment\">   int i = 100;</span></span><br><span class=\"line\"><span class=\"comment\">   int j = i;</span></span><br><span class=\"line\"><span class=\"comment\">   i++;</span></span><br><span class=\"line\"><span class=\"comment\">   return j;</span></span><br><span class=\"line\"><span class=\"comment\">&#125;</span></span><br><span class=\"line\"><span class=\"comment\">*/</span></span><br></pre></td></tr></table></figure>\n<p><strong>final、finally、finalize有什么区别</strong></p>\n<p>final 关键字：final修饰的类无法继承<br />\nfinal修饰的方法无法覆盖<br />\nfinal修饰的变量不能重新赋值。</p>\n<p>finally 关键字：和try一起联合使用，finally语句块中的代码是必须执行的。</p>\n<p>finalize 标识符：是一个Object类中的方法名。这个方法是由垃圾回收器GC负责调用的。</p>\n<p><strong>异常的总结：</strong></p>\n<figure class=\"highlight java\"><table><tr><td class=\"gutter\"><pre><span class=\"line\">1</span><br><span class=\"line\">2</span><br><span class=\"line\">3</span><br><span class=\"line\">4</span><br><span class=\"line\">5</span><br><span class=\"line\">6</span><br><span class=\"line\">7</span><br><span class=\"line\">8</span><br><span class=\"line\">9</span><br><span class=\"line\">10</span><br><span class=\"line\">11</span><br><span class=\"line\">12</span><br><span class=\"line\">13</span><br><span class=\"line\">14</span><br><span class=\"line\">15</span><br><span class=\"line\">16</span><br><span class=\"line\">17</span><br><span class=\"line\">18</span><br><span class=\"line\">19</span><br><span class=\"line\">20</span><br><span class=\"line\">21</span><br><span class=\"line\">22</span><br><span class=\"line\">23</span><br><span class=\"line\">24</span><br><span class=\"line\">25</span><br><span class=\"line\">26</span><br><span class=\"line\">27</span><br><span class=\"line\">28</span><br><span class=\"line\">29</span><br><span class=\"line\">30</span><br><span class=\"line\">31</span><br><span class=\"line\">32</span><br><span class=\"line\">33</span><br><span class=\"line\">34</span><br><span class=\"line\">35</span><br><span class=\"line\">36</span><br><span class=\"line\">37</span><br><span class=\"line\">38</span><br><span class=\"line\">39</span><br><span class=\"line\">40</span><br><span class=\"line\">41</span><br><span class=\"line\">42</span><br><span class=\"line\">43</span><br><span class=\"line\">44</span><br><span class=\"line\">45</span><br><span class=\"line\">46</span><br><span class=\"line\">47</span><br><span class=\"line\">48</span><br><span class=\"line\">49</span><br><span class=\"line\">50</span><br><span class=\"line\">51</span><br><span class=\"line\">52</span><br><span class=\"line\">53</span><br><span class=\"line\">54</span><br><span class=\"line\">55</span><br><span class=\"line\">56</span><br><span class=\"line\">57</span><br><span class=\"line\">58</span><br><span class=\"line\">59</span><br><span class=\"line\">60</span><br><span class=\"line\">61</span><br><span class=\"line\">62</span><br><span class=\"line\">63</span><br><span class=\"line\">64</span><br><span class=\"line\">65</span><br><span class=\"line\">66</span><br><span class=\"line\">67</span><br><span class=\"line\">68</span><br><span class=\"line\">69</span><br><span class=\"line\">70</span><br><span class=\"line\">71</span><br><span class=\"line\">72</span><br><span class=\"line\">73</span><br><span class=\"line\">74</span><br><span class=\"line\">75</span><br><span class=\"line\">76</span><br><span class=\"line\">77</span><br><span class=\"line\">78</span><br><span class=\"line\">79</span><br><span class=\"line\">80</span><br><span class=\"line\">81</span><br><span class=\"line\">82</span><br><span class=\"line\">83</span><br><span class=\"line\">84</span><br><span class=\"line\">85</span><br><span class=\"line\">86</span><br><span class=\"line\">87</span><br><span class=\"line\">88</span><br><span class=\"line\">89</span><br><span class=\"line\">90</span><br><span class=\"line\">91</span><br><span class=\"line\">92</span><br><span class=\"line\">93</span><br><span class=\"line\">94</span><br><span class=\"line\">95</span><br><span class=\"line\">96</span><br><span class=\"line\">97</span><br><span class=\"line\">98</span><br><span class=\"line\">99</span><br><span class=\"line\">100</span><br><span class=\"line\">101</span><br><span class=\"line\">102</span><br><span class=\"line\">103</span><br><span class=\"line\">104</span><br><span class=\"line\">105</span><br><span class=\"line\">106</span><br><span class=\"line\">107</span><br><span class=\"line\">108</span><br><span class=\"line\">109</span><br><span class=\"line\">110</span><br><span class=\"line\">111</span><br><span class=\"line\">112</span><br><span class=\"line\">113</span><br><span class=\"line\">114</span><br><span class=\"line\">115</span><br><span class=\"line\">116</span><br><span class=\"line\">117</span><br><span class=\"line\">118</span><br><span class=\"line\">119</span><br><span class=\"line\">120</span><br><span class=\"line\">121</span><br><span class=\"line\">122</span><br></pre></td><td class=\"code\"><pre><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">   \ttest();<span class=\"comment\">//如果这里不处理的话，下面的代码不会执行</span></span><br><span class=\"line\">   \tSystem.out.println(<span class=\"string\">&quot;不执行&quot;</span>);</span><br><span class=\"line\">   &#125; </span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">test</span><span class=\"params\">()</span> </span>&#123;</span><br><span class=\"line\">     <span class=\"comment\">//这里可以省略throws ArithmeticException，因为属于运行时异常，可以不用处理，自然也就可以省略了，结果一样</span></span><br><span class=\"line\">   \t<span class=\"keyword\">int</span> i = <span class=\"number\">1</span>/<span class=\"number\">0</span>;</span><br><span class=\"line\">    <span class=\"comment\">//java.lang.ArithmeticException: / by zero 这个属于运行时异常，底层会自动生成异常对象，然后抛给调用者</span></span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">-----------------------------------------------------------------------------</span><br><span class=\"line\">运行结果:Exception in thread <span class=\"string\">&quot;main&quot;</span> java.lang.ArithmeticException: / by zero</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">        <span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">            test();</span><br><span class=\"line\">        &#125; <span class=\"keyword\">catch</span> (Exception e) &#123;</span><br><span class=\"line\">            e.printStackTrace();</span><br><span class=\"line\">        &#125;</span><br><span class=\"line\">        System.out.println(<span class=\"string\">&quot;执行&quot;</span>);</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">test</span><span class=\"params\">()</span> </span>&#123;  </span><br><span class=\"line\">   \t<span class=\"keyword\">int</span> i = <span class=\"number\">1</span>/<span class=\"number\">0</span>;</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">-----------------------------------------------------------------------------</span><br><span class=\"line\">运行结果:执行    因为main方法中对异常进行了捕获，所以程序会执行之后的代码。</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">       <span class=\"comment\">//跟普通对象一样,不会影响代码的执行,不管是编译时异常还是运行时异常</span></span><br><span class=\"line\">        NullPointerException nullPointerException = <span class=\"keyword\">new</span> NullPointerException(); </span><br><span class=\"line\">        ClassNotFoundException classNotFoundException = <span class=\"keyword\">new</span> ClassNotFoundException();</span><br><span class=\"line\">        System.out.println(<span class=\"string\">&quot;普通对象&quot;</span>);</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">-----------------------------------------------------------------------------</span><br><span class=\"line\">运行结果:普通对象  （手动用<span class=\"keyword\">new</span>创建的异常对象不会对程序的执行造成影响）</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">        test(); <span class=\"comment\">//如果这里不处理的话，下面的代码不会执行，此时需要使用throws进行异常的抛出或者，使用try对异常进行捕获</span></span><br><span class=\"line\">        System.out.println(<span class=\"string\">&quot;不执行&quot;</span>);</span><br><span class=\"line\">   &#125;  </span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">test</span><span class=\"params\">()</span> </span>&#123;  </span><br><span class=\"line\">       <span class=\"comment\">//这里可以省略throws NullPointerException，因为属于运行时异常，可以不用处理，自然也就可以省略了，结果一样</span></span><br><span class=\"line\">   \t\t<span class=\"keyword\">throw</span> <span class=\"keyword\">new</span> NullPointerException();<span class=\"comment\">//这个属于运行时异常，会手动抛给调用者</span></span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\">运行结果:Exception in thread <span class=\"string\">&quot;main&quot;</span> java.lang.NullPointerException</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">   \t<span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">   \t\ttest(); </span><br><span class=\"line\">\t&#125; <span class=\"keyword\">catch</span> (Exception e) &#123;</span><br><span class=\"line\">\t\t<span class=\"comment\">// <span class=\"doctag\">TODO:</span> handle exception</span></span><br><span class=\"line\">\t&#125;  </span><br><span class=\"line\">   \tSystem.out.println(<span class=\"string\">&quot;执行&quot;</span>); </span><br><span class=\"line\">   &#125;  </span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">test</span><span class=\"params\">()</span> </span>&#123;  <span class=\"comment\">//这里可以省略throws NullPointerException，因为属于运行时异常，可以不用处理，自然也就可以省略了，结果一样</span></span><br><span class=\"line\">   \t<span class=\"keyword\">throw</span> <span class=\"keyword\">new</span> NullPointerException();<span class=\"comment\">//这个属于运行时异常，会手动抛给调用者</span></span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\">运行结果:执行</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">   \t <span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">\t\ttest(); <span class=\"comment\">//这里需要在编写时处理,因为接收到的是编译时异常,要么try---catch处理,要么throws,否则编译不通过.</span></span><br><span class=\"line\">\t&#125; <span class=\"keyword\">catch</span> (ClassNotFoundException e) &#123;</span><br><span class=\"line\">\t\t<span class=\"comment\">// TODO Auto-generated catch block\t\t</span></span><br><span class=\"line\">\t&#125;</span><br><span class=\"line\">   \t System.out.println(<span class=\"string\">&quot;编译时异常&quot;</span>);</span><br><span class=\"line\">   &#125;   </span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">test</span><span class=\"params\">()</span> <span class=\"keyword\">throws</span> ClassNotFoundException </span>&#123;  <span class=\"comment\">//如果采用throws,这里就不可以省略throws ClassNotFoundException，因为属于编译时异常，要处理，自然就不可以省略了</span></span><br><span class=\"line\">   \t<span class=\"keyword\">throw</span> <span class=\"keyword\">new</span> ClassNotFoundException();<span class=\"comment\">//这个属于编译时异常，需要用try---catch处理或者throws</span></span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\">运行结果:编译时异常</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">ExceptionTest</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">   \t test();</span><br><span class=\"line\">   \t System.out.println(<span class=\"string\">&quot;执行&quot;</span>);</span><br><span class=\"line\">   &#125; </span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">test</span><span class=\"params\">()</span>  </span>&#123;  <span class=\"comment\">//如果采用throws,这里就不可以省略throws ClassNotFoundException，因为属于编译时异常，要处理，自然就不可以省略了</span></span><br><span class=\"line\">   \t<span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">\t\t<span class=\"keyword\">throw</span> <span class=\"keyword\">new</span> ClassNotFoundException();</span><br><span class=\"line\">\t&#125; <span class=\"keyword\">catch</span> (ClassNotFoundException e) &#123;</span><br><span class=\"line\">\t\t<span class=\"comment\">// TODO Auto-generated catch block</span></span><br><span class=\"line\">\t\te.printStackTrace();</span><br><span class=\"line\">\t&#125;<span class=\"comment\">//这个属于编译时异常，需要用try---catch处理或者throws</span></span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\">运行结果:java.lang.ClassNotFoundException</span><br><span class=\"line\">\t执行</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\"><span class=\"keyword\">public</span> <span class=\"class\"><span class=\"keyword\">class</span> <span class=\"title\">Test</span> </span>&#123;</span><br><span class=\"line\">   <span class=\"function\"><span class=\"keyword\">public</span> <span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">main</span><span class=\"params\">(String[] args)</span> </span>&#123;</span><br><span class=\"line\">   \t<span class=\"keyword\">try</span> &#123;</span><br><span class=\"line\">   \t\taa();<span class=\"comment\">//这里接受到的是ClassNotFoundException,处于编译时异常，不处理编译器报错</span></span><br><span class=\"line\">   \t\tSystem.out.println(<span class=\"string\">&quot;Hello World1&quot;</span>); <span class=\"comment\">//这里不执行</span></span><br><span class=\"line\">\t&#125; <span class=\"keyword\">catch</span> (NullPointerException e) &#123; <span class=\"comment\">//没处理ClassNotFoundException</span></span><br><span class=\"line\">\t\te.printStackTrace();</span><br><span class=\"line\">\t&#125;</span><br><span class=\"line\">   \tSystem.out.println(<span class=\"string\">&quot;Hello World2&quot;</span>);</span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">  <span class=\"function\"><span class=\"keyword\">static</span> <span class=\"keyword\">void</span> <span class=\"title\">aa</span><span class=\"params\">()</span> <span class=\"keyword\">throws</span> ClassNotFoundException </span>&#123;</span><br><span class=\"line\">   <span class=\"comment\">//这里底层会产生java.lang.ArithmeticException,然后抛给main方法这个调用者</span></span><br><span class=\"line\">   \t<span class=\"keyword\">int</span> i = <span class=\"number\">1</span>/<span class=\"number\">0</span>; </span><br><span class=\"line\">   &#125;</span><br><span class=\"line\">&#125;</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br><span class=\"line\">编译不通过,没运行结果.</span><br><span class=\"line\">----------------------------------------------------------------------------</span><br></pre></td></tr></table></figure>\n",
            "tags": []
        }
    ]
}